Spark 3.5: Implement RewriteTablePath

szehon-ho opened the pull request:
This is the implementation for #10920 (an action to prepare metadata for an Iceberg table for DR copy)

This has been used in production for awhile in our setup, although support for rewrite of V2 position delete is new.  I performed the following cleanups while contributing it.

- Made RewriteTableSparkAction code more functional (avoid using member variable on the action to track state)
- Moved some RewriteTableSparkAction code to core Utll classes to avoid having to make some classes public as was previously done.

RussellSpitzer commented on api/src/main/java/org/apache/iceberg/actions/RewriteTablePath.java:
```diff
@@ -91,9 +92,10 @@ interface Result {
-     * Path to a comma-separated list of source and target paths for all files added to the table
-     * between startVersion and endVersion, including original data files and metadata files
-     * rewritten to staging.
+     * Result file list location. This file contains a 'copy plan', a comma-separated list of all
```
This still feels a little ambiguous. 

Maybe?

A file containing a listing of both original file names and file names under the new prefix, comma separated.

szehon-ho replied:
Put the example in the suggestion and reworded.

flyrain reviewed (commented):
Thanks @szehon-ho for working on it. Left some comments. The major concern is the perf in case of multiple delete files in a manifest file.

szehon-ho commented:
@flyrain thanks for review!  I spent some time cleaning it up.  Now more things are moved to RewriteTablePathsUtil, removing the need to make ManifestLists public.  Also , I moved the rewrite of position deletes into a spark job.  It took me a bit of time to realize, I need to actually rewrite the delete entry's bounds as well (before because I rewrote the position delete file inline, the bounds were correctly populated, but now I have to fix them because the delete file is rewritten later).

szehon-ho commented on core/src/main/java/org/apache/iceberg/util/ContentFileUtil.java:
```diff
@@ -60,28 +69,63 @@ public static CharSequence referencedDataFile(DeleteFile deleteFile) {
+   * @param targetPrefix target prefix which will replace it
+   * @return metrics for the new delete file entry
+   */
+  public static Metrics replacePathBounds(
```
this new logic is to rewrite the position delete entry's file_path bounds, which are used in the delete index.  now we moved the position delete rewrite to a separate spark job instead of inline with writing the delete entry, we need to fix the bounds for the metadata here.

flyrain commented on core/src/main/java/org/apache/iceberg/util/ContentFileUtil.java:
```diff
@@ -60,28 +69,63 @@ public static CharSequence referencedDataFile(DeleteFile deleteFile) {
+      return metricsWithoutPathBounds(deleteFile);
+    }
+
+    if (lowerPathBound.equals(upperPathBound)) {
```
I might miss the context. Do we have to handle the cases that lower path bound doesn't equals to upper path bound? Or are always the same in case of file path in a pos delete file? How does filtering work in that case?

szehon-ho replied:
Yea if I understand it correctly, the existing logic doesnt do filtering if the two bounds arent equal:, see just above this method: 
 https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/util/ContentFileUtil.java#L81

I believe the reasoning is, its not even worth it (as paths are so random).  So for simplicity I just remove it if the two bounds are not the same, what do you think?  Or let me know if you prefer I just change both paths.

szehon-ho replied:
Another context, I think its related to file-scoped delete files, which looks to be the default going forward after this discussion:  https://lists.apache.org/thread/2mpmdp4ns66gp595c9b3clstgskbslsp

flyrain replied:
I think it's worth to have a comment to clarify the behavior. It isn't a blocker to me though.

flyrain reviewed (commented):
Thanks @szehon-ho  for working on it.  LGTM overall. Left questions and minor comments.

dramaticlly commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -0,0 +1,578 @@
+    DataFile newDataFile =
+        DataFiles.builder(spec).copy(entry.file()).withPath(targetDataFilePath).build();
+    appendEntryWithFile(entry, writer, newDataFile);
+    return Pair.of(sourceDataFilePath, newDataFile.location());
```
nit: Looks like writeDataFileEntry and writeDeleteFIle entry returns Pair and RewriteResult respectively, same for all the caller on the stack, I am wondering if we shall use the generic RewriteResult for both?

dramaticlly commented on core/src/main/java/org/apache/iceberg/RewriteTablePathUtil.java:
```diff
@@ -0,0 +1,578 @@
+      throws IOException {
+    String path = deleteFile.location();
+    if (!path.startsWith(sourcePrefix)) {
+      throw new UnsupportedOperationException(
+          "Expected delete file to be under the source prefix: "
+              + sourcePrefix
+              + " but was "
+              + path);
+    }
```
nit, how about `String.format("Expected delete file %s to start with prefix: %s", path, sourcePrefix)`

dramaticlly commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -0,0 +1,720 @@
+        return log.file();
+      }
+    }
+    return null;
```
since this is only used in LINE 210 above and immediate followed by a null check, maybe we can just throw exception here instead?

dramaticlly commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -0,0 +1,720 @@
+
+    // rebuild version files
+    RewriteResult<Snapshot> rewriteVersionResult = rewriteVersionFiles(endMetadata);
+    Set<Snapshot> diffSnapshots =
+        getDiffSnapshotIds(startMetadata, rewriteVersionResult.toRewrite());
```
nit, I think deltaSnapshots might read a bit better

dramaticlly commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -0,0 +1,720 @@
+        return Sets.newHashSet(
+            lastVersionFiles
+                .distinct()
+                .filter(functions.column("added_snapshot_id").isInCollection(diffSnapshotIds))
```
nit, might worth use reference instead of hardcoded string `.filter(functions.column(ManifestFile.SNAPSHOT_ID.name()).isInCollection(diffSnapshotIds))`

dramaticlly commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -0,0 +1,720 @@
+      }
+    } catch (Exception e) {
+      throw new UnsupportedOperationException(
+          "Failed to build the manifest files dataframe, the end version you are "
+              + "trying to copy may contain invalid snapshots, please a younger version that doesn't have invalid "
+              + "snapshots",
```
I think we might miss a verb in second half of the sentence, how about
`"Unable to build the manifest files dataframe. The end version in use may contain invalid snapshots. Please choose an earlier version without invalid snapshots"`

dramaticlly commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -0,0 +1,1003 @@
+            .select("file_path")
+            .as(Encoders.STRING())
+            .collectAsList();
+    assertThat(validDataFilesAfterRebuilt.size()).isEqualTo(2);
+    for (String item : validDataFilesAfterRebuilt) {
+      assertThat(item).startsWith(targetTableLocation);
+    }
```
This can be simplified with 
```java
assertThat(validDataFilesAfterRebuilt)
.hasSize(2)
.allMatch(item -> item.startsWith(targetTableLocation));
```

dramaticlly commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -0,0 +1,1003 @@
+  }
+
+  @Test
+  public void testSameLocations() throws Exception {
```
nit: looks like we dont need `throws Exception` here

dramaticlly commented on spark/v3.5/spark/src/test/java/org/apache/iceberg/spark/actions/TestRewriteTablePathsAction.java:
```diff
@@ -0,0 +1,1003 @@
+    List<ThreeColumnRecord> actualRecords =
+        resultDF.sort("c1", "c2", "c3").as(Encoders.bean(ThreeColumnRecord.class)).collectAsList();
+
+    List<ThreeColumnRecord> expectedRecords = Lists.newArrayList();
+    expectedRecords.add(new ThreeColumnRecord(1, "AAAAAAAAAA", "AAAA"));
+    expectedRecords.add(new ThreeColumnRecord(1, "AAAAAAAAAA", "AAAA"));
```
I am wondering if we can get expectedRecords by reading from source table and compare it against copied table?

dramaticlly reviewed (approved):
thank you @szehon-ho for your great work! I left some nitpicks but overall LGTM

szehon-ho commented:
Rebased and addressed review comments, thanks a lot @flyrain and @dramaticlly for reviewing this big change.

szehon-ho commented on spark/v3.5/spark/src/main/java/org/apache/iceberg/spark/actions/RewriteTablePathSparkAction.java:
```diff
@@ -0,0 +1,731 @@
+    }
+  }
+
+  public static class RewriteContentFileResult extends RewriteResult<ContentFile<?>> {
```
Note: this extra override class is added because Spark encoder didnt like the parent class (RewriteResult) that has type parameter T, hence the need to make a concrete parameterized class.  

It also then needed some extra methods to be able to cleanly aggregate RewriteResult of different types , now with change of https://github.com/apache/iceberg/pull/11555#discussion_r1902167881

szehon-ho commented on core/src/main/java/org/apache/iceberg/io/DeleteSchemaUtil.java:
```diff
@@ -43,4 +43,15 @@ public static Schema pathPosSchema() {
     return rowSchema == null ? pathPosSchema() : pathPosSchema(rowSchema);
   }
+
+  public static Schema posDeleteReadSchema(Schema rowSchema) {
```
Somehow after the rebase this is needed.  

Note that all the readers used now seem to no longer even read the 'row' field https://github.com/apache/iceberg/blob/main/data/src/main/java/org/apache/iceberg/data/BaseDeleteLoader.java#L70. 

But as this is a rewrite, I am taking the strategy of RewritePositionDelete and actually reading this field, but as optional to avoid the assert error if it is not found.  https://github.com/apache/iceberg/blob/main/core/src/main/java/org/apache/iceberg/PositionDeletesTable.java#L118. (the reader there is derived from schema of metadata table PositionDeletesTable).

flyrain replied:
This might not directly related to this PR, but seems like the column `row` should NOT be marked as `required` anywhere.

flyrain commented on core/src/main/java/org/apache/iceberg/io/DeleteSchemaUtil.java:
```diff
@@ -43,4 +43,15 @@ public static Schema pathPosSchema() {
+        MetadataColumns.DELETE_FILE_POS,
+        Types.NestedField.optional(
+            MetadataColumns.DELETE_FILE_ROW_FIELD_ID,
+            "row",
```
could we reuse this constant `DELETE_FILE_ROW_FIELD_NAME`?

flyrain reviewed (approved):
Thanks @szehon-ho ! LGTM.

dramaticlly reviewed (approved):
Thanks Szehon for all the changes and put together RewriteResult interface and RewriteContentFileResult!

szehon-ho commented:
@flyrain addressed comments, and also added a unit test for testing case with 'row' column set on position delete file

flyrain commented:
Thanks @szehon-ho for working on this. Feel free to merge it.

szehon-ho commented:
Thanks a lot @flyrain and @dramaticlly for review, we can continue improving this in follow up prs
